### PR TITLE
Success page

### PR DESCRIPTION
**Pull request by kmagida110**

### CODE REVIEW

File: dashboard/functions/helper.R [hunk 1]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
    reference.vector >= lubridate::floor_date(lubridate::now(), unit = time.unit)
   }
 }
+
+groupAndOrder <- function(df, group.col, data.col, percent = TRUE,top.pct = 1){
```
Comment (comckay) [inline dashboard/functions/helper.R]: spaces after commas

File: dashboard/functions/helper.R [hunk 2]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+  
+  # Returns dataframe grouped by group.col with data.col as a sum and ordered
+  # Return dataframe includes columns named total, cumul and group
+  # df: dataframe
+  # group.col: column that will be grouped along x-axis
+  # data.col: numeric column
+  
+  grouped.df <- df %>% dplyr::group_by_(.dots = group.col) %>% 
+    dplyr::summarise_(total = paste('sum( ',data.col,")")) %>%
+    dplyr::arrange(.,desc(total)) %>% dplyr::rename_(group = paste(group.col))
```
Comment (comckay) [inline dashboard/functions/helper.R]: spaces after commas

File: dashboard/functions/helper.R [hunk 3]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+  
+  
+  # Drop NAs
+  grouped.df <- grouped.df[!is.na(grouped.df$group),]
```
Comment (comckay) [inline dashboard/functions/helper.R]: spaces after commas

File: dashboard/functions/helper.R [hunk 4]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+groupAndOrder <- function(df, group.col, data.col, percent = TRUE,top.pct = 1){
+  
+  # Returns dataframe grouped by group.col with data.col as a sum and ordered
+  # Return dataframe includes columns named total, cumul and group
+  # df: dataframe
+  # group.col: column that will be grouped along x-axis
+  # data.col: numeric column
+  
+  grouped.df <- df %>% dplyr::group_by_(.dots = group.col) %>% 
+    dplyr::summarise_(total = paste('sum( ',data.col,")")) %>%
```
Comment (comckay) [inline dashboard/functions/helper.R]: all pipe operators should be the final thing on a given line

File: dashboard/functions/helper.R (hunk 3 above)
Comment (comckay) [inline dashboard/functions/helper.R]: this should be part of the above pipeline keeping this more concise 
`%>% dplyr::filter(ifelse(is.na(group), F, T))`

File: dashboard/functions/make_plots.R [hunk 5]
```diff
@@ -182,3 +183,43 @@ printGGplotly <- function(plt) {
...
+  # x and y lab: labels for x and y axes
+  # percent: Show metrics as % of total
+  # cumul.line: Show cumulative line
+  
+  
+  # Rename columns
+  grouped.df$group <- grouped.df[[group.col]]
+  grouped.df$total <- grouped.df[[data.col]]
+  grouped.df$cumul <- grouped.df[[cumul.col]]
+
```
Comment (comckay) [inline dashboard/functions/make_plots.R]: use `aes_string` instead

File: dashboard/functions/make_plots.R [hunk 6]
```diff
@@ -182,3 +183,43 @@ printGGplotly <- function(plt) {
...
+  
+  plt <- ggplot(grouped.df, aes(x=group, y = total)) +
+            geom_bar(stat="identity", colour = "black") + 
+            labs(x = paste0(x.lab), title = title, y = y.lab) +
+            expand_limits(y=0) + 
+            theme_bw() + 
+            theme(axis.text.x = element_text(angle = 90, hjust = 1))
+  
+  if(cumul.line){
+    plt <- plt + geom_line(aes(x=group, y=cumul, group = 1), colour = "black") +
```
Comment (comckay) [inline dashboard/functions/make_plots.R]: all monadic operators (the + in this case) should be the final things on a line

File: dashboard/get_data/query.R [hunk 7]
```diff
@@ -117,4 +117,33 @@ ga.sessions %>%
...
+              FROM ga_events
+              WHERE event_category = 'Conversion') AS c
+              ON c.session_id = s.session_id
+              GROUP BY parent_info.content_type, parent_info.title, d.node_id, d.descendant_id, des_info.content_type,
+              des_info.title"
+
+# Get dataframe and fill na conversions and sessions with 0
+dbGetQuery(db.connection, grouped.sessions.query)  %>% 
+  mutate(conversions = ifelse(is.na(conversions), 0, conversions)) %>%
+  mutate(sessions = ifelse(is.na(sessions), 0, sessions)) %>%
```
Comment (comckay) [inline dashboard/get_data/query.R]: These can be one mutate statement separated by a comma avoiding the overhead of a new function call and the passing of data

File: dashboard/get_data/query.R [hunk 8]
```diff
@@ -117,4 +117,33 @@ ga.sessions %>%
...
+  {
+    saveRDS(., "dashboard/data/grouped.sessions.conversions.RDS")
+    grouped.sessions.conversions <<- .
```
Comment (comckay) [inline dashboard/get_data/query.R]: we don't need to assign to global env here

File: dashboard/server.R [hunk 9]
```diff
@@ -1,207 +1,602 @@
...
+  #### MAIN ####
+  output$user.satisfaction.total <- renderValueBox({
+    user.satisfaction = factorPercentage(factor.vec = formstack.master$info_found,
+                                         factor.value = "Yes") %>%
+      prettyPercent(round.n = 1)  # flip to char and ensure % sign is included
+    valueBox(
+      user.satisfaction,  
+      "Approval", icon = icon("thumbs-up", lib = "glyphicon"),
+      color = "yellow"
+    )
```
Comment (comckay) [inline dashboard/server.R]: why are you deleting all of the work I did?

File: dashboard/server.R [hunk 10]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
       autoWidth = TRUE,
       dom = 'tp')
     )
+  output$topic.sessions <- renderPlotly({
+    
+    grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
```
Comment (comckay) [inline dashboard/server.R]: monadic operators terminate lines

File: dashboard/server.R [hunk 11]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+      groupAndOrder(.,group.col = 'parent_title', data.col = 'sessions',percent = input$success.rate.percent, top.pct = .8) %>%
```
Comment (comckay) [inline dashboard/server.R]: spaces after commas

File: dashboard/server.R [hunk 12]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+      buildParetoChart(grouped.df = .,x.lab = 'Topics',y.lab = 'Sessions',
+                       title = "Conversions on Top Topics",cumul.line = TRUE) %>%
```
Comment (comckay) [inline dashboard/server.R]: spaces after commas

File: dashboard/server.R [hunk 13]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+      groupAndOrder(.,group.col = 'parent_title', data.col = 'conversions',percent = input$success.rate.percent, top.pct = .8) %>%
+      buildParetoChart(grouped.df = .,x.lab = 'Topics',y.lab = 'Conversions',
+                       title = "Conversions on Top Topics",cumul.line = TRUE) %>%
+      printGGplotly()
+  })
+  
+  output$topic.conversion.rate <- renderPlotly({
+    
+    plt <- grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
+                plyr::ddply(., c('parent_title'), summarise,
```
Comment (comckay) [inline dashboard/server.R]: use dplyr, it has a summarise function

File: dashboard/server.R [hunk 14]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+    grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
+      groupAndOrder(.,group.col = 'parent_title', data.col = 'sessions',percent = input$success.rate.percent, top.pct = .8) %>%
+      buildParetoChart(grouped.df = .,x.lab = 'Topics',y.lab = 'Sessions',
+                       title = "Conversions on Top Topics",cumul.line = TRUE) %>%
+      printGGplotly()
+  })
+  
+  output$topic.conversions <- renderPlotly({
+    
+    grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
```
Comment (comckay) [inline dashboard/server.R]: monadic operators terminate lines

File: dashboard/server.R [hunk 15]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+      groupAndOrder(.,group.col = 'parent_title', data.col = 'conversions',percent = input$success.rate.percent, top.pct = .8) %>%
```
Comment (comckay) [inline dashboard/server.R]: spaces after commas

File: dashboard/functions/helper.R [hunk 16]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+
+groupAndOrder <- function(df, group.col, data.col, percent = TRUE,top.pct = 1){
+  
+  # Returns dataframe grouped by group.col with data.col as a sum and ordered
+  # Return dataframe includes columns named total, cumul and group
+  # df: dataframe
+  # group.col: column that will be grouped along x-axis
+  # data.col: numeric column
+  
+  grouped.df <- df %>% dplyr::group_by_(.dots = group.col) %>%
```
Comment (comckay) [inline dashboard/functions/helper.R]: use `=` not `<-` in functions

File: dashboard/functions/helper.R (hunk 3 above)
Comment (comckay) [inline dashboard/functions/helper.R]: use `=` not `<-` in functions

File: dashboard/functions/helper.R [hunk 17]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+    dplyr::summarise_(total = paste('sum( ',data.col,")")) %>%
+    dplyr::arrange(.,desc(total)) %>% dplyr::rename_(group = paste(group.col))
+  
+  
+  # Drop NAs
+  grouped.df <- grouped.df[!is.na(grouped.df$group),]
+  
+  
+  # Make cumulative column
+  grouped.df$cumul <- cumsum(grouped.df$total)
```
Comment (comckay) [inline dashboard/functions/helper.R]: use `=` not `<-` in functions

File: dashboard/functions/helper.R [hunk 18]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+  
+  
+  if(percent){
+    data.total <- sum(grouped.df$total)
```
Comment (comckay) [inline dashboard/functions/helper.R]: use `=` not `<-` in functions

File: dashboard/functions/helper.R [hunk 19]
```diff
@@ -87,3 +87,34 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+    grouped.df$total <- grouped.df$total / data.total * 100
+    grouped.df$cumul <- cumsum(grouped.df$total)
+  }
+  
+  grouped.df <- grouped.df[top.pct >= (grouped.df$cumul / sum(grouped.df$total)),]
```
Comment (comckay) [inline dashboard/functions/helper.R]: use `=` not `<-` in functions

File: dashboard/functions/make_plots.R [hunk 20]
```diff
@@ -182,3 +183,43 @@ printGGplotly <- function(plt) {
...
+  # group.col: column that will be grouped along x-axis
+  # data.col: numeric column
+  # cumul.col: cumulative totals
+  # x and y lab: labels for x and y axes
+  # percent: Show metrics as % of total
+  # cumul.line: Show cumulative line
+  
+  
+  # Rename columns
+  grouped.df$group <- grouped.df[[group.col]]
```
Comment (comckay) [inline dashboard/functions/make_plots.R]: use `=` not `<-` in functions

File: dashboard/functions/make_plots.R [hunk 21]
```diff
@@ -182,3 +183,43 @@ printGGplotly <- function(plt) {
...
+  grouped.df$total <- grouped.df[[data.col]]
+  grouped.df$cumul <- grouped.df[[cumul.col]]
+  
+  # Reorder factors largest to smallest
+  grouped.df <- transform(grouped.df, group = reorder(group, order(total, decreasing = TRUE)))
```
Comment (comckay) [inline dashboard/functions/make_plots.R]: use `=` not `<-` in functions

File: dashboard/functions/make_plots.R [hunk 22]
```diff
@@ -182,3 +183,43 @@ printGGplotly <- function(plt) {
...
+  
+  plt <- ggplot(grouped.df, aes(x=group, y = total)) +
```
Comment (comckay) [inline dashboard/functions/make_plots.R]: use `=` not `<-` in functions

File: dashboard/functions/make_plots.R (hunk 6 above)
Comment (comckay) [inline dashboard/functions/make_plots.R]: use `=` not `<-` in functions

File: dashboard/server.R [hunk 23]
```diff
@@ -1,207 +1,602 @@
...
+      geom_bar(stat = "identity") +
+      ggtitle("User Satisfaction Across OS - Top 10") +
+      xlab("") +
+      coord_flip() +
+      ylab("% of Users Satisfied") +
+      theme_bw() 
+    
+      printGGplotly(plt)
+  })  
+
```
Comment (comckay) [inline dashboard/server.R]: Where is all of this coming from?

File: dashboard/server.R [hunk 24]
```diff
@@ -342,4 +737,37 @@ shinyServer(function(input, output) {
...
+    grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
+      groupAndOrder(.,group.col = 'parent_title', data.col = 'conversions',percent = input$success.rate.percent, top.pct = .8) %>%
+      buildParetoChart(grouped.df = .,x.lab = 'Topics',y.lab = 'Conversions',
+                       title = "Conversions on Top Topics",cumul.line = TRUE) %>%
+      printGGplotly()
+  })
+  
+  output$topic.conversion.rate <- renderPlotly({
+    
+    plt <- grouped.sessions.conversions %>% dplyr::filter(., parent_type == 'Topic') %>%
```
Comment (comckay) [inline dashboard/server.R]: use `=` not `<-` in functions

**Comment by kmagida110** [comment]
Really not sure how the server changes are coming in, I think not pulling before pushing does it. I resolved all the diffs, so I'm not sure what dropped.

**Comment by comckay** [comment]
that needs to be resolved before I will merge this

**Comment by kmagida110** [comment]
Cut and pasted from dev with my additions should be up to date.

File: dashboard/get_data/query.R (hunk 8 above)
Comment (kmagida110) [inline dashboard/get_data/query.R]: I thought you did? I was just conforming, will remove

File: dashboard/get_data/query.R (hunk 7 above)
Comment (kmagida110) [inline dashboard/get_data/query.R]: OK

File: dashboard/functions/make_plots.R (hunk 5 above)
Comment (kmagida110) [inline dashboard/functions/make_plots.R]: I would for GGPLOT but need fixed names for transform, leaving as is for now

**Comment by kmagida110** [comment]
OK, should be resolved, double check on server

File: dashboard/server.R (hunk 13 above)
Comment (kmagida110) [inline dashboard/server.R]: IS there a major difference?

File: dashboard/functions/make_plots.R (hunk 5 above)
Comment (kmagida110) [inline dashboard/functions/make_plots.R]: That is a really cool option though, will be using it in the future

File: dashboard/server.R (hunk 13 above)
Comment (comckay) [inline dashboard/server.R]: yes, speed for one

**Comment by comckay** [comment]
server looks good. please clean up according to the above comments and I will merge

File: dashboard/functions/make_plots.R (hunk 5 above)
Comment (comckay) [inline dashboard/functions/make_plots.R]: gotcha that's fine then

File: dashboard/functions/helper.R (hunk 18 above)
Comment (kmagida110) [inline dashboard/functions/helper.R]: Set

File: dashboard/server.R (hunk 12 above)
Comment (kmagida110) [inline dashboard/server.R]: Set

File: dashboard/server.R (hunk 15 above)
Comment (kmagida110) [inline dashboard/server.R]: Set

File: dashboard/server.R (hunk 24 above)
Comment (kmagida110) [inline dashboard/server.R]: Set

File: dashboard/functions/make_plots.R (hunk 21 above)
Comment (kmagida110) [inline dashboard/functions/make_plots.R]: Done

File: dashboard/functions/make_plots.R (hunk 22 above)
Comment (kmagida110) [inline dashboard/functions/make_plots.R]: Done

File: dashboard/functions/make_plots.R (hunk 6 above)
Comment (kmagida110) [inline dashboard/functions/make_plots.R]: Done

Comment (kmagida110) [inline dashboard/functions/make_plots.R]: Done

File: dashboard/get_data/query.R (hunk 7 above)
Comment (kmagida110) [inline dashboard/get_data/query.R]: Resolved

File: dashboard/server.R (hunk 11 above)
Comment (kmagida110) [inline dashboard/server.R]: Done

File: dashboard/server.R (hunk 13 above)
Comment (kmagida110) [inline dashboard/server.R]: Switched

**Comment by kmagida110** [comment]
OK, should be set

File: dashboard/functions/helper.R [hunk 25]
```diff
@@ -87,3 +87,31 @@ flagIncompleteTimeperiod <- function(reference.vector, time.unit) {
...
+  # Returns dataframe grouped by group.col with data.col as a sum and ordered
+  # Return dataframe includes columns named total, cumul and group
+  # df: dataframe
+  # group.col: column that will be grouped along x-axis
+  # data.col: numeric column
+  
+  grouped.df = df %>% 
+    dplyr::group_by_(.dots = group.col) %>% 
+    dplyr::summarise_(total = paste('sum( ', data.col,")")) %>%
+    dplyr::arrange(., desc(total)) %>% dplyr::rename_(group = paste(group.col)) %>%
```
Comment (comckay) [inline dashboard/functions/helper.R]: monadic operators terminate lines

**Comment by comckay** [comment]
from what I see there is still cleanup that needs to occur. There are still `<-` within functions and some `%>%` operators in line for pipelines longer than 2 functions

File: dashboard/functions/helper.R (hunk 25 above)
Comment (kmagida110) [inline dashboard/functions/helper.R]: Did server not helper

**Comment by kmagida110** [comment]
Looks good

**Comment by kmagida110** [comment]
Or at least as far as I can see

**Pull request merged**